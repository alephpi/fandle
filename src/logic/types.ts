export type PinyinStyle = 'plain' | 'strict'
export type MatchType = 'matchAll' | 'match2' | 'match1' | 'miss'

export interface ParsedChar {
  char: string
  _shengmu: string
  _yunmu: string
  _diao: string
  yin: string
}

export interface ParsedPinyin {
  _shengmu: string
  _yunmu: string
  _diao: string
  yin: string
}
export interface MatchResult {
  // char: MatchType
  _shengmu: boolean
  _yunmu: boolean
  _diao: boolean
  yin: MatchType
}

export interface TriesMeta {
  answer?: boolean
  start?: number
  end?: number
  duration?: number
  failed?: boolean
  passed?: boolean
  tries?: string[]
  hint?: boolean
  hintLevel?: number
  strict?: boolean
  sent?: boolean
}

