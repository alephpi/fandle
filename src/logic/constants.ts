export const WORD_LENGTH = 4
export const TRIES_LIMIT = 5
// monthIndex is not month!
export const START_DATE = new Date(2022, 8, 0)
export const RANDOM_SEED = 'fandle'
export const DAYS_PLAY_BACK = 3
// export const DEPLOY_HOST = 'handle.antfu.me'
export const NETLIFY_FUNCTION_HOST = '/.netlify/functions'
