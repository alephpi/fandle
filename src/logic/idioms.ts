import { toSimplified } from '@hankit/tools'
import IdiomPinyinDictRaw from '../data/总成语表.json'

export const IdiomPinyinDict = IdiomPinyinDictRaw as Record<string, string>

export function getIdiom(word: string): [string, string | undefined] | undefined {
  const simplified = toSimplified(word)
  if (IdiomPinyinDict[word])
    return [word, IdiomPinyinDict[word]]
  if (IdiomPinyinDict[simplified])
    return [simplified, IdiomPinyinDict[simplified]]
  return undefined
}

export function getPinyin(word: string) {
  const data = getIdiom(word)
  const parts = data?.[1]
    ? data[1].split(/\s+/g)
    : [] // 在汉兜中，即便某词语不在词库中，输入该词仍然会预览其拼音，故需借助getpinyinraw返回拼音，但反切汉兜直接取消麻烦的拼音显示，故此处不必。为不破坏代码整体，这里仅返回空表。
  // https://baike.baidu.com/item/%E6%B1%89%E8%AF%AD%E6%8B%BC%E9%9F%B3%E6%96%B9%E6%A1%88/1884432
  // 将jqxy后的u还原为v
  return parts.map(i => i
    .replace(/^(y|j|q|x)u([a-z]*[0-9]?)$/g, '$1v$2'),
  )
}
