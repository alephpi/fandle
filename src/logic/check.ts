import PinyinRaw from '../data/answers.txt?raw'
import { WORD_LENGTH } from './constants'
import { getIdiom } from './idioms'

const PinyinList = PinyinRaw.split('\n').map(i => i.trim())
export function filterNonChineseChars(input: string) {
  return Array.from(input)
    .filter(i => /\p{Script=Han}/u.test(i))
    .slice(0, WORD_LENGTH)
    .join('')
}

export function checkValidIdiom(word: string) {
  return !!getIdiom(word)
}

export function checkValidPinyin(pinyin: string) {
  return PinyinList.includes(pinyin)
}
