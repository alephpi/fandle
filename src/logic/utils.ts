import seedrandom from 'seedrandom'
import { pinyinInitials, pinyinInitialsStrict } from '@hankit/tools'
import PinyinFreqPerIdiom from '../data/py_freq_per_idiom.json'
import type { MatchResult, ParsedChar, ParsedPinyin, PinyinStyle } from './types'
import { getPinyin } from './idioms'
import { EXAMPLE_NUMBERS } from './constants'

export function parsePinyin(pinyin: string, pyStyle = 'plain') {
  const tone = pinyin.match(/[\d]$/)?.[0] || ''
  const rawpinyin = pinyin
  if (tone)
    pinyin = pinyin.slice(0, -tone.length).trim()

  let parts: string[] = []

  if (pyStyle === 'plain') {
    if (pinyin) {
      let rest = pinyin
      const one = pinyinInitials.find(i => rest.startsWith(i))
      if (one)
        rest = rest.slice(one.length)
      parts = [one, rest].filter(Boolean) as string[]
    }
  }
  // otherwise equal to strict
  else {
    if (pinyin) {
      let rest = pinyin
      const one = pinyinInitialsStrict.find(i => rest.startsWith(i))
      if (one)
        rest = rest.slice(one.length)
      rest = rest.replace(/yv/, 'v').replace(/yi?/, 'i').replace(/wu?/, 'u').replace(/iu/, 'iou').replace(/ui/, 'uei').replace(/un/, 'uen')
      parts = [one, rest].filter(Boolean) as string[]
    }
  }
  // 现在始终有拼音了，因为我提供了总表

  // if there is no final, actually it's no intital
  if (parts[0] && !parts[1]) {
    parts[1] = parts[0]
    parts[0] = ''
  }

  const [one, two] = parts

  return {
    _shengmu: one,
    _yunmu: two,
    _diao: tone,
    rawpinyin,
  }
}

function parseChar(char: string, pinyin: string, pyStyle?: PinyinStyle): ParsedChar {
  return {
    char,
    parsedPinyin: parsePinyin(pinyin, pyStyle),
  }
}

export function parseWord(word: string, answer?: string, pyStyle?: PinyinStyle) {
  const pinyins = getPinyin(word)
  const chars = Array.from(word)
  // const answerPinyin = answer ? getPinyin(answer) : undefined

  return chars.map((char, i): ParsedChar => {
    // this '' is in case that word is empty (when rendering a new blank word block)
    const pinyin = pinyins[i] || ''
    return parseChar(char, pinyin, pyStyle)
  })
}

export function parseAnswer(answer: string, pyStyle?: PinyinStyle) {
  return parsePinyin(answer, pyStyle)
}

export function testAnswer(input: ParsedChar[], answer: ParsedPinyin) {
  return input.map((char): MatchResult => {
    // const char = toSimplified(a.char)
    const a = char.parsedPinyin
    const shengmu = a._shengmu === answer._shengmu
    const yunmu = a._yunmu === answer._yunmu
    const diao = a._diao === answer._diao
    return {
      _shengmu: shengmu,
      _yunmu: yunmu,
      _diao: diao,
      yin: shengmu && yunmu && diao
        ? 'matchAll'
        : (shengmu && yunmu && !diao) || (shengmu && !yunmu && diao) || (!shengmu && yunmu && diao)
          ? 'match2'
          : (shengmu && !yunmu && !diao) || (!shengmu && yunmu && !diao) || (!shengmu && !yunmu && diao)
            ? 'match1'
            : 'miss',
    }
  })
}

export function checkPass(result: MatchResult[]) {
  // 只要有一个中就通过
  return result.some(r => r.yin === 'matchAll')
}

export function getHint(word: string) {
  // hint is the tone
  return word.slice(-1)
}

export function getAnswerStatistics(answer: keyof typeof PinyinFreqPerIdiom) {
  return PinyinFreqPerIdiom[answer].freq
}

export function getAnswerExamples(answer: keyof typeof PinyinFreqPerIdiom) {
  const examples = PinyinFreqPerIdiom[answer].examples
  return examples.sort(() => 0.5 - Math.random()).slice(0, EXAMPLE_NUMBERS)
}
const numberChar = ['零', '一', '二', '三', '四', '五', '六', '七', '八', '九']
const tens = ['', '十', '百', '千']

export function numberToHanzi(number: number) {
  const digits = Array.from(number.toString()).map(i => +i)
  const chars = digits.map((i, idx) => {
    const unit = i !== 0 ? tens[digits.length - 1 - idx] : ''
    return numberChar[i] + unit
  })
  const str = chars.join('')
  return str
    .replace(/零+/, '零')
    .replace(/(.)零$/, '$1')
}

/**
* Checks whether a given date is in daylight saving time.
* @param date the date object to be checked.
* @returns true if the date is in daylight saving time, false if it's not.
*/
export function isDstObserved(date: Date) {
  const jan = new Date(date.getFullYear(), 0, 1)
  const jul = new Date(date.getFullYear(), 6, 1)
  const standardTimezoneOffset = Math.max(jan.getTimezoneOffset(), jul.getTimezoneOffset())
  return date.getTimezoneOffset() < standardTimezoneOffset
}
