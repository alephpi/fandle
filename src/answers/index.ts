import seedrandom from 'seedrandom'
import type { DictType } from '../logic'
import { getHint } from '../logic'
import { answers as answersMandarin } from './list_mandarin'
import { answers as answersCantonese } from './list_cantonese'
export function getAnswerOfDay(day: number, mode: DictType) {
  let [word, hint] = ['', '']
  const seed = seedrandom(`day-${day}`)()
  if (mode === 'cantonese') {
    if (day >= answersCantonese.length)
      [word = '', hint = ''] = answersCantonese[Math.floor(seed * answersCantonese.length)] || []
    else
      [word = '', hint = ''] = answersCantonese[day] || []
  }
  else {
    if (day >= answersMandarin.length)
      [word = '', hint = ''] = answersMandarin[Math.floor(seed * answersMandarin.length)] || []
    else
      [word = '', hint = ''] = answersMandarin[day] || []
  }
  return {
    word,
    hint: hint || getHint(word),
  }
}
