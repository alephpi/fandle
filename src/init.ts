import { breakingUpdate, initialized, markEnd, markStart, meta, pauseTimer } from './storage'
import { dayNo, daySince, isDev, isFinished, isPassed, showCheatSheet, showHelp } from './state'
import { t } from './i18n'
// import { sendAnalytics } from './analytics'
import { getAnswerOfDay } from './answers'
import { answers as answersMandarin } from './answers/list_mandarin'
import { answers as answersCantonese } from './answers/list_cantonese'
import { START_DATE } from './logic/constants'
import { tryFixAnswer } from './logic/answer-fix'
import type { DictType } from './logic'

useTitle(computed(() => `${t('name')} - ${t('description')}`))

if (!initialized.value)
  showHelp.value = true

watchEffect(() => {
  if (isPassed.value)
    meta.value.passed = true
})

watch(daySince, (n, o) => {
  // on day changed
  if (o === dayNo.value && isFinished.value)
    dayNo.value = n
})

watch([isFinished, meta], () => {
  if (isFinished.value)
    markEnd()
  // sendAnalytics()
}, { flush: 'post' })

watch(isFinished, (v) => {
  if (v)
    showCheatSheet.value = false
}, { flush: 'post' })

const visible = useDocumentVisibility()

let leaveTime = 0
const REFRESH_TIME = 1000 * 60 * 60 * 3 // 3 hours
watchEffect(() => {
  if (visible.value === 'visible') {
    // left for a long while, refresh the page for updates
    if (leaveTime && Date.now() - leaveTime > REFRESH_TIME)
      location.reload()

    // restart timer
    if (meta.value.duration)
      markStart()
  }
  else if (visible.value === 'hidden') {
    leaveTime = Date.now()
    pauseTimer()
  }
}, { flush: 'post' })

nextTick(() => {
  // if (acceptCollecting.value)
  //   sendAnalytics()

  tryFixAnswer(dayNo.value)
})
function showAnswerInfo(mode: DictType) {
  let answerlist: string[][]
  const answer = getAnswerOfDay(dayNo.value, mode)
  if (mode === 'cantonese')
    answerlist = answersCantonese
  else
    answerlist = answersMandarin
  const theDate = new Date(+START_DATE + dayNo.value * 86400000)
  // eslint-disable-next-line no-console
  console.log(`D${dayNo.value}`, theDate.toLocaleDateString(), answer.word, answer.hint)
  // eslint-disable-next-line no-console
  console.log(`${answerlist.length} days prepared`)
  // eslint-disable-next-line no-console
  console.log(`${answerlist.length - dayNo.value} days left`)
  if ((answerlist.length - daySince.value) < 10)
    throw new Error('Not enough days left!')
}

if (isDev || import.meta.hot) {
  if (import.meta.hot) {
    showAnswerInfo('mandarin')
    showAnswerInfo('cantonese')
  }
}

