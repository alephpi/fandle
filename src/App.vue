<script setup lang="ts">
import '~/init'
import { dayNo, daySince } from '~/state'

const { height } = useWindowSize()

watchEffect(() => {
  document.documentElement.style.setProperty('--vh', `${height.value / 100}px`)
})
</script>

<template>
  <main font-sans text="center gray-700 dark:gray-300" select-none>
    <NotTodayBanner v-if="dayNo < daySince" />
    <router-view />
    <Confetti />
  </main>
</template>
