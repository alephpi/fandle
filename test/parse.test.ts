import { expect, it } from 'vitest'
import { parseChar, parsePinyin, parseWord, testAnswer } from '../src/logic'

// it('parseChar', () => {
//   expect(parseChar('虎')).toMatchSnapshot()
//   expect(parseChar('微')).toMatchSnapshot()
//   expect(parseChar('姜')).toMatchSnapshot()
//   expect(parseChar('寿')).toMatchSnapshot()
//   expect(parseChar('翁')).toMatchSnapshot()
// })

// it('testAnswer', () => {
//   expect(testAnswer(parseWord('安'), parseWord('安'))).toMatchSnapshot()
//   expect(testAnswer(parseWord('帆'), parseWord('安'))).toMatchSnapshot()
// })

it('parsePinyin', () => {
  expect(parsePinyin('yi', 'strict')).toMatchSnapshot()
  expect(parsePinyin('yve', 'strict')).toMatchSnapshot()
  expect(parsePinyin('you', 'strict')).toMatchSnapshot()
  expect(parsePinyin('wu', 'strict')).toMatchSnapshot()
  expect(parsePinyin('wei', 'strict')).toMatchSnapshot()
  expect(parsePinyin('dui', 'strict')).toMatchSnapshot()
  expect(parsePinyin('tui', 'strict')).toMatchSnapshot()
  expect(parsePinyin('jvn', 'strict')).toMatchSnapshot()
  expect(parsePinyin('dun', 'strict')).toMatchSnapshot()
})
