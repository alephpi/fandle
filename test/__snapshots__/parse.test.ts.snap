// Vitest Snapshot v1

exports[`parsePinyin 1`] = `
{
  "_diao": "",
  "_shengmu": "",
  "_yunmu": "i",
  "rawpinyin": "yi",
}
`;

exports[`parsePinyin 2`] = `
{
  "_diao": "",
  "_shengmu": "",
  "_yunmu": "ve",
  "rawpinyin": "yve",
}
`;

exports[`parsePinyin 3`] = `
{
  "_diao": "",
  "_shengmu": "",
  "_yunmu": "iou",
  "rawpinyin": "you",
}
`;

exports[`parsePinyin 4`] = `
{
  "_diao": "",
  "_shengmu": "",
  "_yunmu": "u",
  "rawpinyin": "wu",
}
`;

exports[`parsePinyin 5`] = `
{
  "_diao": "",
  "_shengmu": "",
  "_yunmu": "uei",
  "rawpinyin": "wei",
}
`;

exports[`parsePinyin 6`] = `
{
  "_diao": "",
  "_shengmu": "d",
  "_yunmu": "uei",
  "rawpinyin": "dui",
}
`;

exports[`parsePinyin 7`] = `
{
  "_diao": "",
  "_shengmu": "t",
  "_yunmu": "uei",
  "rawpinyin": "tui",
}
`;

exports[`parsePinyin 8`] = `
{
  "_diao": "",
  "_shengmu": "j",
  "_yunmu": "vn",
  "rawpinyin": "jvn",
}
`;

exports[`parsePinyin 9`] = `
{
  "_diao": "",
  "_shengmu": "d",
  "_yunmu": "uen",
  "rawpinyin": "dun",
}
`;
